### PR TITLE
Better error reporting.

### PR DESCRIPTION
**Pull request by PrimusLV**

### CODE REVIEW

File: src/legionpe/theta/utils/ReportErrorTask.php [hunk 1]
```diff
@@ -38,19 +38,27 @@ public function onRun(){
...
+		$message .= "Server: ".Settings::$LOCALIZE_IP . " " . Settings::$LOCALIZE_PORT." (".Settings::$LOCALIZE_ID.")".PHP_EOL;
+		$message .= "=== END ERROR ===".PHP_EOL;
+		self::log($message);
 		self::$last0 = self::$last1;
 		self::$last1 = self::$last2;
 		self::$last2 = self::$last3;
 		self::$last3 = microtime(true);
 	}
 	public static function log($line){
 		echo $line, PHP_EOL;
```
Comment (PrimusLV) [inline src/legionpe/theta/utils/ReportErrorTask.php]: I don't see any use in this

**Comment by SOF3** [comment]
@PrimusLV improve your wiki edit. We could even use it to create an automatic installer ;)

**Comment by SOF3** [comment]
@PrimusLV but why use `TEXT` instead of `VARCHAR(32767)`?

And, skin1 and skin2 are binary according to https://github.com/LegionPE/LegionPE-Theta-Base/blob/master/src/legionpe/theta/query/SaveSinglePlayerQuery.php#L66

File: src/legionpe/theta/utils/ReportErrorTask.php [hunk 2]
```diff
@@ -38,19 +38,27 @@ public function onRun(){
...
-		self::log("ping-all !!! PEMapModder ping!");
-		self::log("Exception caught: " . $this->exMsg);
-		self::log("In file: " . $this->exFile . "#" . $this->exLine);
-		self::log("Happened during: " . $this->when);
-		self::log("On the server below:");
-		Utils::getURL(Credentials::IRC_WEBHOOK_NOPREFIX . urlencode("BotsHateNames: status " . Settings::$LOCALIZE_IP . " " . Settings::$LOCALIZE_PORT));
+		$message = "=== ERROR BEGIN ===".PHP_EOL;
+		$message .= "Exception: " . $this->exMsg.PHP_EOL;
+		$message .= "File: " . $this->exFile . "#" . $this->exLine.PHP_EOL;
+		$message .= "UNIX Timestamp: " . $this->when.PHP_EOL;
```
Comment (SOF3) [inline src/legionpe/theta/utils/ReportErrorTask.php]: `$this->when` is a string condition, such as `"Executing command /hub"`.

File: src/legionpe/theta/utils/ReportErrorTask.php [hunk 3]
```diff
@@ -38,19 +38,27 @@ public function onRun(){
...
+		$message .= "Time: ".date("H:i:s", $this->when).PHP_EOL;
```
Comment (SOF3) [inline src/legionpe/theta/utils/ReportErrorTask.php]: So remove this line as well.

File: src/legionpe/theta/utils/ReportErrorTask.php [hunk 4]
```diff
@@ -38,19 +38,27 @@ public function onRun(){
 		if(microtime(true) - self::$last0 < 60){
 			return;
 		}
-		self::log("ping-all !!! PEMapModder ping!");
```
Comment (SOF3) [inline src/legionpe/theta/utils/ReportErrorTask.php:41]: Well xD This was supposed to notify PEMapModder on IRC, but this won't happen anymore ;)

File: src/legionpe/theta/utils/ReportErrorTask.php [hunk 5]
```diff
@@ -38,19 +38,27 @@ public function onRun(){
...
-		self::log("Exception caught: " . $this->exMsg);
-		self::log("In file: " . $this->exFile . "#" . $this->exLine);
-		self::log("Happened during: " . $this->when);
-		self::log("On the server below:");
-		Utils::getURL(Credentials::IRC_WEBHOOK_NOPREFIX . urlencode("BotsHateNames: status " . Settings::$LOCALIZE_IP . " " . Settings::$LOCALIZE_PORT));
+		$message = "=== ERROR BEGIN ===".PHP_EOL;
```
Comment (SOF3) [inline src/legionpe/theta/utils/ReportErrorTask.php]: I doubt if you can use PHP_EOL with @TkTech's notifico directly. IRC messages are supposed to be one-liners.

File: src/legionpe/theta/utils/ReportErrorTask.php [hunk 6]
```diff
@@ -38,19 +38,27 @@ public function onRun(){
 		if(microtime(true) - self::$last0 < 60){
 			return;
 		}
-		self::log("ping-all !!! PEMapModder ping!");
-		self::log("Exception caught: " . $this->exMsg);
-		self::log("In file: " . $this->exFile . "#" . $this->exLine);
-		self::log("Happened during: " . $this->when);
-		self::log("On the server below:");
-		Utils::getURL(Credentials::IRC_WEBHOOK_NOPREFIX . urlencode("BotsHateNames: status " . Settings::$LOCALIZE_IP . " " . Settings::$LOCALIZE_PORT));
```
Comment (SOF3) [inline src/legionpe/theta/utils/ReportErrorTask.php:46]: @BotsHateNames was supposed to be a bot that shows this server's status ;)

**Review by SOF3 — CHANGES_REQUESTED** [review]
Check out how this class was constructed first ;)

File: src/legionpe/theta/utils/ReportErrorTask.php [hunk 7]
```diff
@@ -54,6 +54,11 @@ public function onRun(){
 	public static function log($line){
 		echo $line, PHP_EOL;
 		Utils::getURL(Credentials::IRC_WEBHOOK_NOPREFIX . urlencode($line));
-		Utils::postURL(Credentials::SLACK_WEBHOOK, ["payload" => json_encode(["text" => $line, "icon_url" => Credentials::LEGIONPE_ICON_URL, "username" => "Server-".Settings::$LOCALIZE_ID.""])]);
+		Utils::postURL(Credentials::SLACK_WEBHOOK, ["payload" => json_encode([
+			"text" => $line,
+			"icon_url" => Credentials::LEGIONPE_ICON_URL,
+			"username" => "Server-".Settings::$LOCALIZE_ID."",
+			"channel" => "error"])
+			]);
```
Comment (SOF3) [inline src/legionpe/theta/utils/ReportErrorTask.php]: But we don't have such channel on Slack :P

File: src/legionpe/theta/utils/ReportErrorTask.php (hunk 5 above)
Comment (TkTech) [reply src/legionpe/theta/utils/ReportErrorTask.php]: If you're using plain-text hooks it'll just be converted into one message per newline, subject to the normal rate limiting.

Comment (SOF3) [reply src/legionpe/theta/utils/ReportErrorTask.php]: @TkTech thanks!

Still, I suppose it is not necessary to use `PHP_EOL` over `\n`.

File: src/legionpe/theta/utils/ReportErrorTask.php (hunk 7 above)
Comment (PrimusLV) [reply src/legionpe/theta/utils/ReportErrorTask.php]: Shame on you @PrimusLV :D

File: src/legionpe/theta/utils/ReportErrorTask.php (hunk 4 above)
Comment (PrimusLV) [reply src/legionpe/theta/utils/ReportErrorTask.php:41]: Indeed :(

File: src/legionpe/theta/utils/ReportErrorTask.php (hunk 6 above)
Comment (PrimusLV) [reply src/legionpe/theta/utils/ReportErrorTask.php:46]: I'll restore it right away

File: src/legionpe/theta/utils/ReportErrorTask.php [hunk 8]
```diff
@@ -38,19 +38,21 @@ public function onRun(){
...
-		self::log("On the server below:");
-		Utils::getURL(Credentials::IRC_WEBHOOK_NOPREFIX . urlencode("BotsHateNames: status " . Settings::$LOCALIZE_IP . " " . Settings::$LOCALIZE_PORT));
+		$lines = ["=== ERROR BEGIN ==="];
+		$lines[] = "Exception: " . $this->exMsg;
+		$lines[] = "File: " . $this->exFile . "#" . $this->exLine;
+		$lines[] = "When: " . $this->when;
+		$lines[] = "Time: ".date("H:i:s", time());
+		$lines[] = "Server: ".Settings::$LOCALIZE_IP . " " . Settings::$LOCALIZE_PORT." (".Settings::$LOCALIZE_ID.")";
+		$lines[] = "=== END ERROR ===";
+		foreach($lines as $line) self::log($line);
```
Comment (PrimusLV) [inline src/legionpe/theta/utils/ReportErrorTask.php:48]: Sending single lines without line breaks.

**Review by SOF3 — APPROVED** [review]

**Pull request merged**